### ISSUE TITLE
SIDBlaster-USB

### ISSUE THREAD

**Issue by gh0stless** [jpage8580/GTUltra#46]
Interested in better SIDBlaster USB support? I would provide two kits or devices to the developers for free. The hardsid.dll was already ported for Linux and Mac by Ken Handel.

**Comment by acrouzet** [comment]
Is SIDBlaster-USB even supported? I tried putting the hardsid.dll into the win32 binary folder and setting Hardsid device number to 1 in the .cfg but the sound init failed.

**Comment by marcinkubica** [comment]
@gh0stless yes, insterested. I also have HardSID Uno I have been using back in the days and would like it to work under my mac. Any more info on the port please? (github repo?) Much appreciated.

**Comment by gh0stless** [comment]
https://github.com/gh0stless/SIDBlasterUSB_HardSID-emulation-driver/

**Comment by gh0stless** [comment]
Hi @marcinkubica, I made some fixes and added a description document on how to use the library, with example code. Hopefully this now helps with POSIX support for GTUltra.

Note: hooking up a SIDBlaster on Linux with D2XX is a bit tricky, so keep that in mind — it should really be handled by a proper installer.

**Comment by gh0stless** [comment]
@marcinkubica I am currently creating the new binaries for Linux and Mac OS, please be patient.

**Comment by gh0stless** [comment]
@marcinkubica the new binaries are up: https://github.com/gh0stless/SIDBlasterUSB_HardSID-emulation-driver/releases/tag/v516

- Windows, Linux (x64), and macOS (universal - Intel + Apple Silicon, so it should also work on your HardSID Uno's host if that's Apple Silicon) plain binaries, no installer needed.
- Linux note: SIDBlaster-USB needs its D2XX driver to actually claim the device instead of the kernel's generic serial (VCP) driver - the archive's `libhardsid.so` alone isn't enough, you also need the udev rule that detaches `ftdi_sio`/`usbserial` on attach. Full installer with that handled automatically is on crazy-midi.de if you don't want to do it by hand.
- Also includes SIDBlasterTool (device config utility - read serial/SID type, program them) for all three platforms.

Tested on real SIDBlaster-USB hardware on Linux and macOS myself, sounds identical to the Windows result. Let me know if it works for you.

**Comment by gh0stless** [comment]
Follow-up for whoever wants to actually wire up Linux/macOS SIDBlaster-USB support in GTUltra - looked at `src/gsound.c` and it's less work than it might look like.

The SID-write protocol/timing logic is already correct and platform-agnostic - the whole sound loop (`SIDWRITEDELAY`/`SIDWAVEDELAY`, flush handling, etc. around line ~394-521) calls through function pointers (`HardSID_Write`, `HardSID_Delay`, `HardSID_Flush`, `HardSID_SoftFlush`, `MuteHardSID_Line`, ...) that exactly match the classic HardSID DLL API. Nothing there needs to change.

What's platform-specific is only how those function pointers get populated:

- **Windows** (line ~868-885): `LoadLibrary("HARDSID.DLL")` + `GetProcAddress` per function - this already works today and needs no changes.
- **Linux/Unix** (line ~77-79, 156-169, 321-332, 505-521): a completely different, older code path - opens `/dev/sid0`/`/dev/sid1` character devices and writes raw packed words via `write()`. That's the old Linux kernel-module HardSID driver model (see `cwsid.h`), not a library call at all, and it's why SIDBlaster-USB (which ships a userspace `.so`/`.dylib`, not a kernel module) doesn't work through this path.

The fix is to add a Unix branch that mirrors the Windows loading block almost 1:1, swapping `LoadLibrary`/`GetProcAddress` for `dlopen`/`dlsym` against `libhardsid.so` (Linux) / `libhardsid.dylib` (macOS), then let the existing `HardSID_Write`/`HardSID_Delay`/`HardSID_Flush`/`HardSID_SoftFlush` calls run unchanged - same shape as the `sidblaster-builder` dlopen loader I wrote for libsidplayfp: https://github.com/gh0stless/libsidplayfp/blob/master/src/builders/sidblaster-builder/sidblaster-builder.cpp

Happy to send a PR for the `gsound.c` dlopen branch if that's useful - let me know.

**Comment by bufka** [comment]
I have ported js lib for HardSID Uno in the new editor. Don't have old hardsid cards anymore (ISA ones?) can test mac only tho.
